File: 剑指offer/链表/24.反转链表/24.反转链表.py
from util.leetcode_type import ListNode

"""
递归
不进入递归中，只思考每次中的效果
递归结束条件：not cur and not cur.next
递归过程：
- 进入递归，把最后一个节点（当前的队尾，结果的队首）传递出去 last = self.reverse(cur.next)
- 当前节点下个的指向当前节点 cur.next.next = cur
- 取消之前指向 cur.next = None
"""


class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        if not head:
            return head
        return self.reverse(head)

    def reverse(self, cur):
        if not cur or not cur.next:
            return cur
        last = self.reverse(cur.next)
        cur.next.next = cur
        cur.next = None
        return last

File: 剑指offer/链表/24.反转链表/24.反转链表.go
package 链表

import "algorithms-code-note/util/structures"

type ListNode = structures.ListNode

/*
迭代方式
pre 前节点，迭代结束后成为队尾节点
cur 当前节点
循环时：
    把cur.next保存到tmp
    把cur.next 只想pre，实现反转
    推进节点-注意推进顺序，先pre后cur：
        - pre = cur 前节点推荐到当前
        - cur = tmp 当前节点推进到原next节点

*/
func reverseList(head *ListNode) *ListNode {
	if head == nil {
		return head
	}
	var pre *ListNode
	cur := head
	for cur != nil {
		tmp := cur.Next
		cur.Next = pre
		pre = cur
		cur = tmp
	}
	return pre
}
